### PR TITLE
feat(net): add peer_id/ip blacklist

### PR DESCRIPTION
**Pull request by Will-Smith11**
Thought I would go ahead and add the blacklist mentioned in  #327, this is just a first revision. before I go any father and write some tests for this. 

Would like to confirm that they way that this is currently built aligns with the wanted design for this.  @mattsse @gakonst thoughts?

### CODE REVIEW

**Comment by Will-Smith11** [comment]
Thought I would go ahead and add the blacklist mentioned in  #327, this is just a first revision. before I go any father and write some tests for this. Would like to confirm that they way that this is currently built aligns with the wanted design for this.

**Review by gakonst — COMMENTED** [review]
i think directionally correct - pending @mattsse comments

File: crates/net/network/src/session/mod.rs
Comment (Rjected) [inline crates/net/network/src/session/mod.rs:644]: Should the checks for blacklisted peer id / IP should happen here, rather than after performing the `p2p` handshake? We know the `remote_addr` here, so we can disconnect early if the IP is blacklisted.

In the case of an outgoing connection: We know the remote peer id, so we can immediately disconnect if it is blacklisted.
In the case of an incoming connection: We can perform the ECIES handshake and disconnect if the `stream.remote_id()` is blacklisted.

**Review by Rjected — CHANGES_REQUESTED** [review]
cool! A test would also be nice, but will defer to other opinions if a test should be left to a followup PR

File: crates/net/network/src/session/mod.rs
Comment (Will-Smith11) [reply crates/net/network/src/session/mod.rs:644]: Ideally at the root for every case would be the most efficient. for outgoing, you could block it right at discovery table updates. and you could ip block on binding. However I decided to not due this to keep the code simple as doing filtering on the same kind of case in multiple places imo is a pretty bad practice because it can cause some nice footguns. I'm happy to change it if the consensus is to take a different approach

**Comment by Will-Smith11** [comment]
> cool! A test would also be nice, but will defer to other opinions if a test should be left to a followup PR

was going to wait for #352  to be merged before adding tests

File: crates/net/network/src/session/mod.rs
Comment (Will-Smith11) [reply crates/net/network/src/session/mod.rs:665]: yeah seems more efficient this way and where not spreading out the filtering all over the place

File: crates/net/network/src/session/mod.rs
Comment (mattsse) [inline crates/net/network/src/session/mod.rs:214]: hmm, I don't really like this here.

instead of checking the inclusion in the list during authentication, I'd rather do this when the pending session task resolves.

Also imo the right place for this check would be the `PeerManager` which is notified for new incoming connections.

**Review by mattsse — CHANGES_REQUESTED** [review]
tysm for picking this up.

this looks pretty good, I have some architectural change requests though.
Moving the `BanList` type to `PeerManager` would also make it easier to add new peers/ips to the list dynamically, no need to Arc and share it then. So we can add new command variants to `Peers|NetworkHandle`

File: crates/net/network/src/session/config.rs
Comment (gakonst) [reply crates/net/network/src/session/config.rs]: yeah and maybe let's always us banlist and ban instead of blacklist order to be sensitive with language, appreciate your understanding there

File: crates/net/network/src/session/mod.rs
Comment (gakonst) [reply crates/net/network/src/session/mod.rs]: We should add some guidelines on what should go where, eg this peer manager recommendation was not obvious to me

File: crates/net/network/src/peers/manager.rs
Comment (gakonst) [inline crates/net/network/src/peers/manager.rs:493]: We probably also want a simple is_ip_banned and is_peerid_banned for future library users that may know one but not the other, but lgtm here

**Review by gakonst — CHANGES_REQUESTED** [review]
Looking good. Can we get a unit test for this case?

**Review by mattsse — CHANGES_REQUESTED** [review]
+1 on change requests @gakonst raised

also, there is `PeersManager::on_inbound_pending_session()` that now needs to be modified as well:
* accept IpAddr
* add new error type with variants: `ExceedsLimit, IpBanned`

**Comment by Will-Smith11** [comment]
@gakonst @mattsse is the plan to leave it up to the user to define how they want to supply the blacklist. from a user perspective it could be nice to be able to define a file that we read on startup and then through the console you can blacklist ips and peer ids that get written to the specified file on shutdown?

File: crates/net/network/src/peers/manager.rs
Comment (gakonst) [inline crates/net/network/src/peers/manager.rs:531]: This should have a thiserror::Error derive

File: crates/net/network/src/peers/manager.rs
Comment (gakonst) [inline crates/net/network/src/peers/manager.rs:569]: Can you assert on the exact error? You can do match -> assert on the right case, and panic on all other cases

File: crates/net/network/src/peers/manager.rs
Comment (gakonst) [inline crates/net/network/src/peers/manager.rs:582]: Love let some syntax

**Review by gakonst — APPROVED** [review]
Awesome work. Nits mainly and defer to Matt otherwise.

File: crates/net/network/tests/it/connect.rs
Comment (gakonst) [inline crates/net/network/tests/it/connect.rs:99]: @Rjected should we add a helper fn which instantiates Geth with a random datadir?

File: crates/net/network/tests/it/connect.rs
Comment (gakonst) [inline crates/net/network/tests/it/connect.rs:105]: @Rjected should we have a helper for returning PeerId here so that consumers don't need to know the internals of enr encode uncompressed etc?

File: crates/net/network/tests/it/connect.rs
Comment (gakonst) [inline crates/net/network/tests/it/connect.rs:122]: Let's add a helper for this too somewhere

File: crates/net/network/tests/it/connect.rs
Comment (gakonst) [inline crates/net/network/tests/it/connect.rs:129]: Should we be doing the banlist check before even the session is opened so we don't need to open and then immediately close? Or is that not possible? @mattsse

File: crates/net/network/src/swarm.rs
Comment (gakonst) [inline crates/net/network/src/swarm.rs:190]: Shouldn't we be propagating the error? This seems to eat it

**Review by gakonst — CHANGES_REQUESTED** [review]

File: crates/net/network/tests/it/connect.rs
Comment (Will-Smith11) [reply crates/net/network/tests/it/connect.rs:129]: for the case with incoming banned  `PeerIds` afaik we need to handshake them to see there id. all other cases we close before opening a session

File: crates/net/network/src/swarm.rs
Comment (Will-Smith11) [reply crates/net/network/src/swarm.rs:190]: I looked at doing this but it didn't make sense to me as I don't think we should be putting out a `NetworkEvent` every time we don't open a connection because an address was blacklisted. I don't think there is anything useful that can be done with this information, so propagating this into a swarm event would just cause the trace to be done at the network manager which seemed unnecessary.

File: crates/net/network/tests/it/connect.rs
Comment (Rjected) [reply crates/net/network/tests/it/connect.rs:99]: Yeah we should, we will probably need to do it for every geth instance in a test.

File: crates/net/network/tests/it/connect.rs
Comment (Rjected) [reply crates/net/network/tests/it/connect.rs:105]: Yep, will make an issue for adding enode utilities, also generally documenting and providing helper functions so consumers don't need to interact with raw keys, compressed vs uncompressed, etc. Consumers should not need to know about EC point encoding to use the p2p APIs.

Currently is a sore spot that I think could lead to api misuse and confusion, so I can put more effort on this

Comment (Rjected) [reply crates/net/network/tests/it/connect.rs:105]: added in #381

File: crates/net/network/tests/it/connect.rs
Comment (Rjected) [reply crates/net/network/tests/it/connect.rs:122]: made an issue for it: #382

File: crates/net/network/tests/it/connect.rs
Comment (gakonst) [reply crates/net/network/tests/it/connect.rs:129]: thanks - makes sense!

File: crates/net/network/tests/it/connect.rs
Comment (gakonst) [reply crates/net/network/tests/it/connect.rs:105]: Agreed - thanks!

File: crates/net/network/src/swarm.rs
Comment (gakonst) [reply crates/net/network/src/swarm.rs:190]: got it - makes sense unless @mattsse disagrees

**Comment by gakonst** [comment]
This LGTM - thanks for taking care of all the nits! If @mattsse approves we can merge.

**Comment by gakonst** [comment]
This LGTM - thanks for taking care of all the nits! If @mattsse approves we can merge.

**Review by mattsse — APPROVED** [review]
lgtm, tysm

**Comment by mattsse** [comment]
test currently fail, I think we just need to install geth in CI

ref https://github.com/gakonst/ethers-rs/blob/master/.github/workflows/ci.yml#L37

**Comment by gakonst** [comment]
@Rjected did in separate PR I think

**Comment by gakonst** [comment]
https://github.com/paradigmxyz/reth/pull/373

**Pull request merged**
